klippa/src/serialize.rs

```diff
@@ -303,6 +454,86 @@ impl Serializer {
+        }
+
+        if self.current.is_some() || self.packed.is_empty() {
+            self.set_err(SerializeErrorFlags::SERIALIZE_ERROR_OTHER);
+            return;
+        }
+
+        let mut offset_links = Vec::new();
+        for parent_obj_idx in self.packed.iter() {
+            let Some(parent_obj) = self.object_pool.get_obj(*parent_obj_idx) else {
+                self.set_err(SerializeErrorFlags::SERIALIZE_ERROR_OTHER);
+                return;
+            };
+
+            for link in parent_obj.real_links.iter() {
+                let Some(child_pool_idx) = self.packed.get(link.objidx) else {
+                    self.set_err(SerializeErrorFlags::SERIALIZE_ERROR_OTHER);
+                    return;
+                };
+                let Some(child_obj) = self.object_pool.get_obj(*child_pool_idx) else {
+                    self.set_err(SerializeErrorFlags::SERIALIZE_ERROR_OTHER);
+                    return;
+                };
+
+                let offset = match link.whence {
+                    OffsetWhence::Head => child_obj.head - parent_obj.head,
+                    OffsetWhence::Tail => child_obj.head - parent_obj.tail,
+                    OffsetWhence::Absolute => self.head - self.start + child_obj.head - self.tail,
+                };
+
+                if offset < link.bias as usize {
+                    self.set_err(SerializeErrorFlags::SERIALIZE_ERROR_OTHER);
+                    return;
+                }
+
+                let offset = offset - link.bias as usize;
+                let offset_start_pos = parent_obj.head + link.position as usize;
+                offset_links.push((offset_start_pos, link.width, offset));
+            }
+        }
+
+        for (offset_start_pos, offset_width, offset) in offset_links.iter() {
+            self.assign_offset(*offset_start_pos, *offset_width, *offset);
+        }
+    }
+
+    //TODO: take care of signed offset
+    fn assign_offset(&mut self, offset_start_pos: usize, link_width: LinkWidth, offset: usize) {
+        let (offset_width, max_offset) = match link_width {
+            LinkWidth::Two => (2, u16::MAX as usize),
+            LinkWidth::Three => (3, (Uint24::MAX).to_u32() as usize),
+            LinkWidth::Four => (4, u32::MAX as usize),
+        };
+
+        if offset > max_offset {
+            self.set_err(SerializeErrorFlags::SERIALIZE_ERROR_OFFSET_OVERFLOW);
+            return;
+        }
+
+        let Some(offset_data_bytes) = self
+            .data
+            .get_mut(offset_start_pos..offset_start_pos + offset_width)
+        else {
+            self.set_err(SerializeErrorFlags::SERIALIZE_ERROR_OTHER);
+            return;
+        };
+
+        let be_bytes = (offset as u32).to_be_bytes();
+        match link_width {
+            LinkWidth::Two => offset_data_bytes.copy_from_slice(&[be_bytes[2], be_bytes[3]]),
+            LinkWidth::Three => {
+                offset_data_bytes.copy_from_slice(&[be_bytes[1], be_bytes[2], be_bytes[3]])
```

Comment:
could maybe just be,
```suggestion
            LinkWidth::Two => offset_data_bytes.copy_from_slice(&be_bytes[2..=3]),
            LinkWidth::Three => {
                offset_data_bytes.copy_from_slice(&be_bytes[1..=3])
```